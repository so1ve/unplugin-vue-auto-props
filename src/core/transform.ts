import MagicString from "magic-string";

import {
	findDefineComponentCall,
	haveDefineComponentImport,
	parse,
} from "./utils";

const generateDefineProps = (name: string, props: string[]) =>
	`\nObject.defineProperty(${name}, "props", {
  value: ${JSON.stringify(props)},
});`;

export function transform(code: string) {
	const parsed = parse(code);
	if (!haveDefineComponentImport(parsed)) {
		return;
	}
	const magicString = new MagicString(code);
	const defineComponentCalls = findDefineComponentCall(parsed);
	for (const [name, props, start] of defineComponentCalls) {
		magicString.appendRight(start + 1, generateDefineProps(name, props));
	}

	return { code: magicString.toString(), map: magicString.generateMap() };
}
