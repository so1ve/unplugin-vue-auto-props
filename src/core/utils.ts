import type { ParseResult } from "@babel/parser";
import { parse as babelParse } from "@babel/parser";
import * as t from "@babel/types";
import _traverse from "@babel/traverse";

const traverse =
  typeof (_traverse as any).default === "undefined"
    ? _traverse
    : ((_traverse as any).default as typeof _traverse);

type Parsed = ParseResult<t.File>;

export const parse = (code: string) =>
  babelParse(code, {
    sourceType: "module",
    plugins: ["typescript", "jsx"],
  });

const getKeyName = (node: any) => node.key.name;

/** Determine if the code has a `defineComponent` import from `vue` */
export function haveDefineComponentImport(parsed: Parsed) {
  let haveDefineComponentValueImport = false;
  traverse(parsed, {
    ImportDeclaration({ node }) {
      if (node.source.value === "vue") {
        for (const specifier of node.specifiers) {
          if (
            t.isImportSpecifier(specifier) &&
            "name" in specifier.imported &&
            specifier.imported.name === "defineComponent"
          ) {
            haveDefineComponentValueImport = true;
          }
        }
      }
    },
  });

  return haveDefineComponentValueImport;
}

/** Get the property signatures from type definition node */
function findTypeDefinitionMembersByName(
  parsed: Parsed,
  name: string,
): t.TSPropertySignature[] | null {
  let result: t.TSPropertySignature[] | null = null;
  traverse(parsed, {
    TSInterfaceDeclaration({ node }) {
      if (node.id.name === name) {
        result = node.body.body as any;
      }
    },
    TSTypeAliasDeclaration({ node }) {
      if (node.id.name === name) {
        result = (node.typeAnnotation as any).members;
      }
    },
  });

  return result;
}

function resolveTypeDefinitionByTypeAnnotation(
  parsed: Parsed,
  typeAnnotation: t.TSTypeAnnotation | t.TSTypeReference | t.TSTypeLiteral,
): string[] {
  if ("typeName" in typeAnnotation && "name" in typeAnnotation.typeName) {
    const { name: typeName } = typeAnnotation.typeName;
    const typeDefinitionMembers = findTypeDefinitionMembersByName(
      parsed,
      typeName,
    );

    return typeDefinitionMembers?.map(getKeyName) ?? [];
  } else if (t.isTSTypeLiteral(typeAnnotation)) {
    return typeAnnotation.members.map(getKeyName);
  }

  return [];
}

const isAllowedTypeAnnotation = (
  typeAnnotation: any,
): typeAnnotation is t.TSTypeLiteral | t.TSTypeReference =>
  t.isTSTypeLiteral(typeAnnotation) ||
  (t.isTSTypeReference(typeAnnotation) &&
    "typeName" in typeAnnotation &&
    "name" in typeAnnotation.typeName);

/** Get member keys from a CallExpression, extracts its typeDefinition */
function findPropTypeMemberKeys(parsed: Parsed, node: t.CallExpression) {
  if (node.typeParameters) {
    const [param] = node.typeParameters.params;
    if (t.isTSTypeLiteral(param)) {
      return param.members.map(getKeyName);
    }
    if (!t.isTSTypeReference(param) || !("name" in param.typeName)) {
      return [];
    }
    const { name: typeName } = param.typeName;
    const typeDefinitionMembers = findTypeDefinitionMembersByName(
      parsed,
      typeName,
    );

    return typeDefinitionMembers?.map(getKeyName) ?? [];
  }

  const [arg] = node.arguments;

  if (t.isObjectExpression(arg)) {
    const properties = arg.properties.filter(
      (p) => t.isObjectProperty(p) || t.isObjectMethod(p),
    );
    const propsPropetry = properties.find(
      (p) => t.isObjectProperty(p) && "name" in p.key && p.key.name === "props",
    );
    // It already has a props property, so we don't need to add it
    if (propsPropetry) {
      return [];
    }
    const setupProperty = properties.find(
      (p) =>
        (t.isObjectMethod(p) || t.isObjectProperty(p)) &&
        "name" in p.key &&
        p.key.name === "setup",
    );

    if (
      !setupProperty ||
      !(t.isObjectMethod(setupProperty) || t.isObjectProperty(setupProperty))
    ) {
      return [];
    }
    const setup = (
      t.isObjectMethod(setupProperty) ? setupProperty : setupProperty.value
    ) as t.ObjectMethod | t.ArrowFunctionExpression;

    const [propsParam] = setup.params;

    const typeAnnotation =
      "typeAnnotation" in propsParam &&
      propsParam.typeAnnotation &&
      "typeAnnotation" in propsParam.typeAnnotation &&
      propsParam.typeAnnotation?.typeAnnotation;

    if (!typeAnnotation || !isAllowedTypeAnnotation(typeAnnotation)) {
      return [];
    }

    return resolveTypeDefinitionByTypeAnnotation(parsed, typeAnnotation);
  }

  if (t.isArrowFunctionExpression(arg) || t.isFunctionExpression(arg)) {
    const propsParam = arg.params[0];
    if (
      propsParam?.typeAnnotation &&
      "typeAnnotation" in propsParam.typeAnnotation &&
      isAllowedTypeAnnotation(propsParam.typeAnnotation.typeAnnotation)
    ) {
      return resolveTypeDefinitionByTypeAnnotation(
        parsed,
        propsParam.typeAnnotation.typeAnnotation,
      );
    }
  }

  return [];
}

export function findDefineComponentCall(
  parsed: Parsed,
): [componentName: string, props: string[], end: number][] {
  const result: FindDefineComponentCallReturn = [];
  traverse(parsed, {
    VariableDeclarator({ node: { id, init, end } }) {
      if (
        t.isCallExpression(init) &&
        t.isIdentifier(init?.callee) &&
        t.isIdentifier(id) &&
        init.callee.name === "defineComponent"
      ) {
        const memberKeys = findPropTypeMemberKeys(parsed, init);
        if (memberKeys.length > 0) {
          result.push([id.name, memberKeys, end!]);
        }
      }
    },
  });

  return result;
}
