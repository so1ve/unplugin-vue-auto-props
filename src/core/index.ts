import { createUnplugin } from "unplugin";

import { transform } from "./transform";
import type { Options } from "./types";

export default createUnplugin<Options | undefined>((_options) => ({
	name: "unplugin-vue-tsx-auto-props",
	enforce: "pre",
	transform(code, id) {
		if (!id.endsWith(".tsx")) {
			return;
		}

		return transform(code);
	},
}));
