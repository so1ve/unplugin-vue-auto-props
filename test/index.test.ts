import { describe, expect, it } from "vitest";

import { transform } from "../src/core/transform";

describe("should", () => {
  it("transform", () => {
    const code = `
import { defineComponent } from "vue";

interface Props<T> {
  a: number;
  b: {
    some: boolean
  }
}

type P1 = {
  a: 1
}

const Foo = defineComponent(
  <T>(props: Props<T>) => {
    return () => <div>{props}</div>;
  },
);

const Bar = defineComponent(
  <T>(props: P1<T>) => {
    return () => <div>{props}</div>;
  },
);
`;
    expect(transform(code)).toMatchInlineSnapshot(`
      {
        "code": "
      import { defineComponent } from \\"vue\\";

      interface Props<T> {
        a: number;
        b: {
          some: boolean
        }
      }

      type P1 = {
        a: 1
      }

      const Foo = defineComponent(
        <T>(props: Props<T>) => {
          return () => <div>{props}</div>;
        },
      );
      Object.defineProperty(Foo, \\"props\\", {
        value: [\\"a\\",\\"b\\"],
      });

      const Bar = defineComponent(
        <T>(props: P1<T>) => {
          return () => <div>{props}</div>;
        },
      );
      Object.defineProperty(Bar, \\"props\\", {
        value: [\\"a\\"],
      });
      ",
        "map": SourceMap {
          "file": undefined,
          "mappings": "AAAA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;;;GAAE;AACF;AACA;AACA;AACA;AACA;AACA;;;GAAE;",
          "names": [],
          "sources": [
            "",
          ],
          "sourcesContent": undefined,
          "version": 3,
        },
      }
    `);
  });

  it("should not transform default export", () => {
    const code = `
import { defineComponent } from "vue";

interface Props<T> {
  a: number;
  b: {
    some: boolean
  }
}
export default defineComponent(
  <T>(props: Props<T>) => {
    return () => <div>{props}</div>;
  },
);
`;
    expect(transform(code)).toMatchInlineSnapshot(`
      {
        "code": "
      import { defineComponent } from \\"vue\\";

      interface Props<T> {
        a: number;
        b: {
          some: boolean
        }
      }
      export default defineComponent(
        <T>(props: Props<T>) => {
          return () => <div>{props}</div>;
        },
      );
      ",
        "map": SourceMap {
          "file": undefined,
          "mappings": "AAAA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;",
          "names": [],
          "sources": [
            "",
          ],
          "sourcesContent": undefined,
          "version": 3,
        },
      }
    `);
  });

  it("should not transform non-functional components", () => {
    const code = `
import { defineComponent } from "vue";

type Props = {
  foo: number
}

const Foo = defineComponent({
  setup(props: Props) {
    return () => <div>{props}</div>;
  },
});`;
    expect(transform(code)).toMatchInlineSnapshot(`
      {
        "code": "
      import { defineComponent } from \\"vue\\";

      type Props = {
        foo: number
      }

      const Foo = defineComponent({
        setup(props: Props) {
          return () => <div>{props}</div>;
        },
      });",
        "map": SourceMap {
          "file": undefined,
          "mappings": "AAAA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA",
          "names": [],
          "sources": [
            "",
          ],
          "sourcesContent": undefined,
          "version": 3,
        },
      }
    `);
  });

  it("should not transform complex type", () => {
    const code = `
import { defineComponent } from "vue";

type Props = 1 extends 1 ? {
  foo: number
} : { a: string }

const Foo = defineComponent((props: Props) => () => <div>{props}></div>);
`;
    expect(transform(code)).toMatchInlineSnapshot(`
      {
        "code": "
      import { defineComponent } from \\"vue\\";

      type Props = 1 extends 1 ? {
        foo: number
      } : { a: string }

      const Foo = defineComponent((props: Props) => () => <div>{props}></div>);
      ",
        "map": SourceMap {
          "file": undefined,
          "mappings": "AAAA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;",
          "names": [],
          "sources": [
            "",
          ],
          "sourcesContent": undefined,
          "version": 3,
        },
      }
    `);
  });

  it("should not transform without type annotation", () => {
    const code = `
import { defineComponent } from "vue";

const Foo = defineComponent((props) => () => <div>{props}></div>);
`;

    expect(transform(code)).toMatchInlineSnapshot(`
      {
        "code": "
      import { defineComponent } from \\"vue\\";

      const Foo = defineComponent((props) => () => <div>{props}></div>);
      ",
        "map": SourceMap {
          "file": undefined,
          "mappings": "AAAA;AACA;AACA;AACA;",
          "names": [],
          "sources": [
            "",
          ],
          "sourcesContent": undefined,
          "version": 3,
        },
      }
    `);
  });

  it("should transform function declaration", () => {
    const code = `
import { defineComponent } from "vue";

type Props = {
  foo: number
}

const Foo = defineComponent(function (props: Props) { return () => <div>{props}></div> });
`;

    expect(transform(code)).toMatchInlineSnapshot(`
      {
        "code": "
      import { defineComponent } from \\"vue\\";

      type Props = {
        foo: number
      }

      const Foo = defineComponent(function (props: Props) { return () => <div>{props}></div> });
      Object.defineProperty(Foo, \\"props\\", {
        value: [\\"foo\\"],
      });
      ",
        "map": SourceMap {
          "file": undefined,
          "mappings": "AAAA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;;;GAA0F;",
          "names": [],
          "sources": [
            "",
          ],
          "sourcesContent": undefined,
          "version": 3,
        },
      }
    `);
  });

  it("should transform type argument", () => {
    const code = `
import { defineComponent } from "vue";

type Props = {
  foo: number
}

const Foo = defineComponent<Props>((props) => () => <div>{props}></div>);
`;

    expect(transform(code)).toMatchInlineSnapshot(`
      {
        "code": "
      import { defineComponent } from \\"vue\\";

      type Props = {
        foo: number
      }

      const Foo = defineComponent<Props>((props) => () => <div>{props}></div>);
      Object.defineProperty(Foo, \\"props\\", {
        value: [\\"foo\\"],
      });
      ",
        "map": SourceMap {
          "file": undefined,
          "mappings": "AAAA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;;;GAAyE;",
          "names": [],
          "sources": [
            "",
          ],
          "sourcesContent": undefined,
          "version": 3,
        },
      }
    `);
  });
});
