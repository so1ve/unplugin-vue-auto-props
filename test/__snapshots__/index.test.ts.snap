// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should > should not transform complex type 1`] = `
{
  "code": "
import { defineComponent } from \\"vue\\";

type Props = 1 extends 1 ? {
  foo: number
} : { a: string }

const Foo = defineComponent((props: Props) => () => <div>{props}></div>);
",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;",
    "names": [],
    "sources": [
      "",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;

exports[`should > should not transform default export 1`] = `
{
  "code": "
import { defineComponent } from \\"vue\\";

interface Props<T> {
  a: number;
  b: {
    some: boolean
  }
}
export default defineComponent(
  <T>(props: Props<T>) => {
    return () => <div>{props}</div>;
  },
);
",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;",
    "names": [],
    "sources": [
      "",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;

exports[`should > should not transform setup option when props is set 1`] = `
{
  "code": "
import { defineComponent } from \\"vue\\";

const Foo = defineComponent({
  props: {
    foo: Number
  },
  setup: (props: { foo: number }) => () => <div>{props}></div>
});
",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;",
    "names": [],
    "sources": [
      "",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;

exports[`should > should not transform without type annotation 1`] = `
{
  "code": "
import { defineComponent } from \\"vue\\";

const Foo = defineComponent((props) => () => <div>{props}></div>);
",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA;AACA;AACA;AACA;",
    "names": [],
    "sources": [
      "",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;

exports[`should > should transform function declaration 1`] = `
{
  "code": "
import { defineComponent } from \\"vue\\";

type Props = {
  foo: number
}

const Foo = defineComponent(function (props: Props) { return () => <div>{props}></div> });
Object.defineProperty(Foo, \\"props\\", {
  value: [\\"foo\\"],
});
",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;;;GAA0F;",
    "names": [],
    "sources": [
      "",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;

exports[`should > should transform inline argument type 1`] = `
{
  "code": "
import { defineComponent } from \\"vue\\";

const Foo = defineComponent((props: { foo: number }) => () => <div>{props}></div>);
Object.defineProperty(Foo, \\"props\\", {
  value: [\\"foo\\"],
});
",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA;AACA;AACA;AACA;;;GAAmF;",
    "names": [],
    "sources": [
      "",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;

exports[`should > should transform inline type 1`] = `
{
  "code": "
import { defineComponent } from \\"vue\\";

const Foo = defineComponent<{ foo: number }>((props) => () => <div>{props}></div>);
Object.defineProperty(Foo, \\"props\\", {
  value: [\\"foo\\"],
});
",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA;AACA;AACA;AACA;;;GAAmF;",
    "names": [],
    "sources": [
      "",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;

exports[`should > should transform setup option 1`] = `
{
  "code": "
import { defineComponent } from \\"vue\\";

const Foo = defineComponent({
  setup: (props: { foo: number }) => () => <div>{props}></div>
});
Object.defineProperty(Foo, \\"props\\", {
  value: [\\"foo\\"],
});
",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA;AACA;AACA;AACA;AACA;AACA;;;GAAG;",
    "names": [],
    "sources": [
      "",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;

exports[`should > should transform type argument 1`] = `
{
  "code": "
import { defineComponent } from \\"vue\\";

type Props = {
  foo: number
}

const Foo = defineComponent<Props>((props) => () => <div>{props}></div>);
Object.defineProperty(Foo, \\"props\\", {
  value: [\\"foo\\"],
});
",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;;;GAAyE;",
    "names": [],
    "sources": [
      "",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;

exports[`should > transform 1`] = `
{
  "code": "
import { defineComponent } from \\"vue\\";

interface Props<T> {
  a: number;
  b: {
    some: boolean
  }
}

type P1 = {
  a: 1
}

const Foo = defineComponent(
  <T>(props: Props<T>) => {
    return () => <div>{props}</div>;
  },
);
Object.defineProperty(Foo, \\"props\\", {
  value: [\\"a\\",\\"b\\"],
});

const Bar = defineComponent(
  <T>(props: P1<T>) => {
    return () => <div>{props}</div>;
  },
);
Object.defineProperty(Bar, \\"props\\", {
  value: [\\"a\\"],
});
",
  "map": SourceMap {
    "file": undefined,
    "mappings": "AAAA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;AACA;;;GAAE;AACF;AACA;AACA;AACA;AACA;AACA;;;GAAE;",
    "names": [],
    "sources": [
      "",
    ],
    "sourcesContent": undefined,
    "version": 3,
  },
}
`;
